### ISSUE TITLE
there are multiple problems with TransformControls

### ISSUE THREAD

**Issue by makc** [mrdoob/three.js#9312]
The effect of scaling is not independent of scene / object dimensions. I had to change [these lines](https://github.com/mrdoob/three.js/blob/master/examples/js/controls/TransformControls.js#L983-L985) to

``` javascript
                        if ( scope.axis === "X" ) scope.object.scale.x = oldScale.x * ( 1 + 0.01 * point.x / oldScale.x );
                        if ( scope.axis === "Y" ) scope.object.scale.y = oldScale.y * ( 1 + 0.01 * point.y / oldScale.y );
                        if ( scope.axis === "Z" ) scope.object.scale.z = oldScale.z * ( 1 + 0.01 * point.z / oldScale.z );

```

to get sane scaling sensitivity in my project quickly.

**Comment by makc** [comment]
[Here is the fiddle](http://jsfiddle.net/hfj7gm6t/1334/). Try to scale the box, I dare you :)

**Comment by makc** [comment]
Oh :shit: there are more problems. 2 out of 3 rotations are broken when the object is edge-on: [fiddle](http://jsfiddle.net/hfj7gm6t/1335/). If it's moved even slightly, these rotations behave much better.

**Comment by makc** [comment]
God damn it! Just found another issue with [translation](http://jsfiddle.net/hfj7gm6t/1336/): you can't move the box using central yellow dot until it is moved around in some other way. And occasionally trying to do so results in the box jumping closer to camera.

**Comment by makc** [comment]
On the bright side, next fork of that fiddle will be 1337

**Comment by makc** [comment]
Aaaaand, [it's mine](http://jsfiddle.net/hfj7gm6t/1337/) again - just realized one can scale by middle point in scaling gizmo, and by "can" I mean "can not". Well, you sort of can, but it is also insane magnitude.

**Comment by makc** [comment]
And how do you call [this](https://github.com/mrdoob/three.js/blob/master/examples/js/controls/TransformControls.js#L750)?? Setting space to "local" if scaling, means never going back to "world" until the user actually looks into source :S

**Comment by makc** [comment]
> 2 out of 3 rotations are broken when the object is edge-on

this is caused by degenerate case of
```javascript
var planeIntersect = intersectObjects( pointer, [ _gizmo[ _mode ].activePlane ] );
```
I could not come up with any kind of clever hack for this - it looks like this needs to be fixed either deep in THREE.Mesh raycasting code, or replace intersectObjects with some sort of analytical intersection for this gizmo.

I tried this stupid hack:
```javascript
		function hackIntersectionPoint (pointer) {
			var ap = _gizmo[_mode].activePlane, hack = false;
			if (_mode == "rotate") {
				var ax = new THREE.Vector3, ay = new THREE.Vector3, az = new THREE.Vector3;
				ap.matrixWorld.extractBasis (ax, ay, az);
				if (
					((ap.name == "YZ") && (ax.y == 0) && (ax.z == 0)) ||
					((ap.name == "XZ") && (ay.x == 0) && (ay.z == 0))
					) {
						ap = _gizmo[_mode].planes.XYZE; hack = true;
					}
			}
			var planeIntersect = intersectObjects( pointer, [ ap ] );
			if (hack) planeIntersect.point.z += 10;
			return planeIntersect;
```
coded specifically for my setup (camera at 000, mesh at 00negativez) - essentially, intersecting with the plane parallel to the screen - while it sort of works it causes different problems, so not a good way to develop more general hack.

**Comment by makc** [comment]
actually, what might work is replacing single planes with two planes at ~1° to each other, so there would be an intersection with one of them

**Comment by arodic** [comment]
@makc the problem is... the central yellow dot is actually the z-facing blue handle looking directly towards the camera. Ideally, the camera-facing handles should be hidden as soon as they are more or less aligned with eye vector. This is because there is no suitable plane to raycast against. So... when X axis is facing camera, only Y and Z should be available and so on.

**Comment by arodic** [comment]
This issue is fixed in alternate version of [TransformControls](https://github.com/arodic/ree.js/blob/master/app/js/controls/Transform.js) But as mentioned in #7947 this fix depends on somewhat modular approach to making big and complex scene manipulation tools.

I hope to revisit this project in near future but I don't see the clear path towards contributing fixes back into threejs. I could manually combine all the dependencies into single monolithic TransformControls.js once I get to it.

**Comment by killroy42** [comment]
I think I hit that scope local issue, where everything seems fine, and suddenly the controls align to local scope.

**Comment by makc** [comment]
@arodic
> I could manually combine all the dependencies into single monolithic TransformControls.js once I get to it.
did you ever get to do it?

**Comment by sandromartis** [comment]
I'm currently also having the problem with the `TransformControls` not doing anything at all when I try to scale in `world` space.
It seems that the code does not support that. It even forces the space back into `local` when switching the mode to `scale`.
Not sure what the reason is for that.

Our issue is that we would like to do transforms on multiple objects. We basically put several objects into a `THREE.Group` and then do the transform on that group.

In that case the gizmos look weird in `local` space if the individual objects have been rotated before hand and already have a rotation applied to them. That's why we would like to just do the scaling in `world` mode, ignoring any rotation and just align the gizmos with the x,y,z axes.
(Does this make sense? 😄 )

Shouldn't it be possible to just do same thing when scaling no matter if space is `local` or `world`? Or does this have some unintended side effect?

Basically removing line 977:
```
if ( scope.space === "local" ) {
``` 

Maybe I have time next week to look at this.

**Comment by makc** [comment]
@arodic I am bumping this with yet another issue. If you omit 2nd argument it subscribes directly to the document, ending up with
> TypeError: domElement.getBoundingClientRect is not a function

Not to mention there is no way to change domElement, meaning you cannot create TransformControls before the canvas.

**Comment by arodic** [comment]
Sorry @makc I haven't done anything with three.js in a long time although I would really love to work on it again.

@sandromartis world space scale is not really possible without another transformation node in the scenegraph. I would suggest to parent the object you want to world-scale to an empty Object3D, and scale that instead. If you look at how Maya handles that for example, scaling in world space, doesen't even modify objects scale attribute. Instead, scaling in world space modifies vertex positions in world space and its not something you cant reset/undo to neutral scale.

**Comment by arodic** [comment]
Hey @makc you should check out my pull request. A bunch of these issues should be fixed now. If you can help testing, it would be great!

**Comment by makc** [comment]
@arodic damn, that was so long ago. I can only suggest you to compare [the fiddle](http://jsfiddle.net/hfj7gm6t/1334/) I posted with [its fork using new version](http://jsfiddle.net/hfj7gm6t/5062/).

**Comment by makc** [comment]
oh, there were more than 1 fiddles!

**Comment by makc** [comment]
> just realized one can scale by middle point in scaling gizmo, and by "can" I mean "can not". Well, you sort of can, but it is also insane magnitude.

this middle point no longer exists.

**Comment by makc** [comment]
> Setting space to "local" if scaling, means never going back to "world" until the user actually looks into source :S
no fiddle for this, so I have to trust you it's fixed :P

**Comment by arodic** [comment]
Hmm I'm seeing some weird issues in the fiddle that I didn't see in the editor and TransformControl example. For example, scaling to 0 breaks the controls completely. see [here](http://jsfiddle.net/hfj7gm6t/5064/) Also rotations are not quite right. I'll need to investigate this.

>this middle point no longer exists.

Yeah, I moved the uniform scale handles to the ends so I can have some distance from center to do relative scale from.

>no fiddle for this, so I have to trust you it's fixed :P

Yes, the scale mode is still only local space but it wont switch other modes to local.

**Comment by kavikode** [comment]
when there are multiple viewports on one canvas, transformcontrols only is controlled by only the center viewport. I'd like for each of viewport to handle each XYZ plane. As of now is not possible to this is a problem that needs to be fixed.
